### PR TITLE
Add version requirements for deploying

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Changes

Adds support for version requirements which are checked when deploying PostHog.

@guidoiaquinti @tiina303 @fuziontech we talked about this at standup and this also aligns very well with my big migrations work.

**Note:**

_Why is this a class and not some bespoke method?_

Because I'll be using this for the big migrations work as well. Having it generalizable makes it easier to reuse in different places, add to, and test.

_Why do we not throw if the version is invalid?_

I wanted to start with a soft warning that can be overridden until we try this in production for a bit and introduce hard requirements.

## How did you test this code?

Added tests for the `VersionRequirement` class.

### CODE REVIEW

File: posthog/settings.py [hunk 1]
```diff
@@ -731,3 +732,10 @@ def show_toolbar(request):
 
 # keep in sync with plugin-server
 EVENTS_DEAD_LETTER_QUEUE_STATSD_METRIC = "events_added_to_dead_letter_queue"
+
+VERSION_REQUIREMENTS = [VersionRequirement(service="postgres", version_range=">=14.0.0,<15.0.0",)]
```
Comment (yakkomajuri) [inline posthog/settings.py:736]: @guidoiaquinti @tiina303 what do we want to have here? (reminder that these are soft requirements for now)

File: posthog/settings.py [hunk 2]
```diff
@@ -731,3 +732,10 @@ def show_toolbar(request):
...
+
+if PRIMARY_DB == AnalyticsDBMS.CLICKHOUSE:
+    VERSION_REQUIREMENTS = VERSION_REQUIREMENTS + [
+        VersionRequirement(service="clickhouse", version_range=">=21.6.0,<21.7.0"),
```
Comment (yakkomajuri) [inline posthog/settings.py:740]: @guidoiaquinti @tiina303 and what do we want here?

**Comment by yakkomajuri** [comment]
hahaha this works too well!

GH Actions:
Screenshot 2021-11-23 at 10 03 05

File: posthog/apps.py [hunk 3]
```diff
@@ -32,3 +33,13 @@ def ready(self):
             posthoganalytics.disabled = True
         elif settings.TEST or os.environ.get("OPT_OUT_CAPTURE", False):
             posthoganalytics.disabled = True
+
+        for version_requirement in settings.VERSION_REQUIREMENTS:
+            [in_range, version] = version_requirement.is_service_in_accepted_version()
+            if not in_range:
+                start_anyway = input(
```
Comment (guidoiaquinti) [inline posthog/apps.py:40]: is this something we would like to support? If yes, we should also add a `--skip-version-requirements` or `--force` flag to skip this check entirely

File: posthog/apps.py [hunk 4]
```diff
@@ -32,3 +33,13 @@ def ready(self):
             posthoganalytics.disabled = True
         elif settings.TEST or os.environ.get("OPT_OUT_CAPTURE", False):
             posthoganalytics.disabled = True
+
+        for version_requirement in settings.VERSION_REQUIREMENTS:
```
Comment (guidoiaquinti) [inline posthog/apps.py:37]: As  `version_requirement` is actually a service, should we rename it?

File: posthog/settings.py (hunk 1 above)
Comment (guidoiaquinti) [inline posthog/settings.py]: nit: `PostgreSQL`

File: posthog/apps.py (hunk 4 above)
Comment (yakkomajuri) [reply posthog/apps.py:37]: well `version_requirement` is an instance of `VersionRequirement`

File: posthog/apps.py (hunk 3 above)
Comment (yakkomajuri) [reply posthog/apps.py:40]: Not in the future, no. But as a first step, yes. I just want to see how this behaves out in the wild before being strict.

From the PR description:

> Why do we not throw if the version is invalid?
> 
> I wanted to start with a soft warning that can be overridden until we try this in production for a bit and introduce hard requirements.


Maybe we could send a PostHog event to get a sense of people being out of range 🤔

File: posthog/settings.py (hunk 1 above)
Comment (guidoiaquinti) [reply posthog/settings.py]: I think version range is 👍 

Nit: `version_range` -> `supported_version` (as technically we could also support a single version and not a range)

File: posthog/version_requirement.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
+from semantic_version.base import SimpleSpec, Version
+
+
+class VersionRequirement:
+    accepted_services = ("clickhouse", "postgres")
```
Comment (guidoiaquinti) [inline posthog/version_requirement.py:5]: Ditto

File: posthog/settings.py (hunk 1 above)
Comment (yakkomajuri) [reply posthog/settings.py]: Yeah I was in the dilemma of "version" not covering when we use ranges, and "range" not covering when we use a specific version.

Will update though.

File: posthog/version_requirement.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+            return self.get_postgres_version()
+
+        if self.service == "clickhouse":
+            return self.get_clickhouse_version()
+
+    def get_postgres_version(self):
+        from django.db import connection
+
+        with connection.cursor() as cursor:
+            cursor.execute("SHOW server_version")
```
Comment (guidoiaquinti) [inline posthog/version_requirement.py:38]: Did we verify pgbouncer doesn't intercept and bypass this query returning some custom output? I don't expect it but still better to verify it

File: posthog/version_requirement.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    def get_clickhouse_version(self):
+        from ee.clickhouse.client import sync_execute
+
+        rows = sync_execute("SELECT version()")
+        version = rows[0][0]
+
+        return self.version_string_to_semver(version)
+
+    @staticmethod
+    def version_string_to_semver(version):
```
Comment (guidoiaquinti) [inline posthog/version_requirement.py:54]: Shouldn't we use something like:

```
>>> import semantic_version
>>> v = semantic_version.Version('0.1.1')
>>> v.major
0
```
?

File: posthog/apps.py (hunk 4 above)
Comment (guidoiaquinti) [reply posthog/apps.py:37]: Yep but each item is technically a _service version requirement_. I find the wording a bit confusing but feel free to skip this suggestion 👍

File: posthog/version_requirement.py (hunk 7 above)
Comment (yakkomajuri) [reply posthog/version_requirement.py:54]: That's what I had, but that won't parse a lot of stuff we want to parse. 

e.g. my local Postgres is on version `14.0`, which will throw here.

ClickHouse also has 4 levels in their versioning, which isn't supported either.

Hence the parsing. We lose a bit of granularity, but it's enough for us.

File: posthog/settings.py [hunk 8]
```diff
@@ -731,3 +732,10 @@ def show_toolbar(request):
 
 # keep in sync with plugin-server
 EVENTS_DEAD_LETTER_QUEUE_STATSD_METRIC = "events_added_to_dead_letter_queue"
+
+VERSION_REQUIREMENTS = [VersionRequirement(service="postgresql", supported_version=">=12.0.0,<15.0.0",)]
```
Comment (guidoiaquinti) [inline posthog/settings.py:736]: I think we need 11 to 13 here or we'll break cloud (see [this](https://github.com/PostHog/posthog/issues/6624))

File: posthog/version_requirement.py (hunk 7 above)
Comment (yakkomajuri) [reply posthog/version_requirement.py:54]: I've added more handling. There are a few more edge cases, so this custom handling actually makes some sense.

e.g. `11.13 (Ubuntu 11.13-2.heroku1+1)`

File: ee/apps.py [hunk 9]
```diff
@@ -1,6 +1,19 @@
 from django.apps import AppConfig
 
+from ee.settings import VERSION_REQUIREMENTS_EE
+
 
 class EnterpriseConfig(AppConfig):
     name = "ee"
     verbose_name = "Enterprise"
+
+    def ready():
```
Comment (guidoiaquinti) [inline ee/apps.py:10]: `check_prerequisites()`?  Then we can technically add multiple checks (like `check_service_version_requirement`...)

File: posthog/settings.py (hunk 8 above)
Comment (yakkomajuri) [reply posthog/settings.py]: Yeah just checked this

File: ee/settings.py [hunk 10]
```diff
@@ -85,3 +86,7 @@
 MATERIALIZE_COLUMNS_BACKFILL_PERIOD_DAYS = get_from_env("MATERIALIZE_COLUMNS_BACKFILL_PERIOD_DAYS", 90, type_cast=int)
 # Maximum number of columns to materialize at once. Avoids running into resource bottlenecks (storage + ingest + backfilling).
 MATERIALIZE_COLUMNS_MAX_AT_ONCE = get_from_env("MATERIALIZE_COLUMNS_MAX_AT_ONCE", 10, type_cast=int)
+
+VERSION_REQUIREMENTS_EE = [
+    VersionRequirement(service="clickhouse", supported_version=">=21.6.0,<21.7.0"),
```
Comment (guidoiaquinti) [inline ee/settings.py:91]: Should we add `Kafka` for the ee version?

File: posthog/settings.py [hunk 11]
```diff
@@ -733,9 +733,4 @@ def show_toolbar(request):
 # keep in sync with plugin-server
 EVENTS_DEAD_LETTER_QUEUE_STATSD_METRIC = "events_added_to_dead_letter_queue"
 
-VERSION_REQUIREMENTS = [VersionRequirement(service="postgresql", supported_version=">=12.0.0,<15.0.0",)]
-
-if PRIMARY_DB == AnalyticsDBMS.CLICKHOUSE:
-    VERSION_REQUIREMENTS = VERSION_REQUIREMENTS + [
-        VersionRequirement(service="clickhouse", supported_version=">=21.6.0,<21.7.0"),
-    ]
+VERSION_REQUIREMENTS = [VersionRequirement(service="postgresql", supported_version=">=12.0.0,<15.0.0")]
```
Comment (guidoiaquinti) [inline posthog/settings.py:736]: Should we add Redis for the default version?

File: ee/apps.py (hunk 9 above)
Comment (yakkomajuri) [reply ee/apps.py:10]: `ready` is a Django hook that will be called when the app is set up and ready

Comment (yakkomajuri) [reply ee/apps.py]: anyway, removed from here

File: posthog/version_requirement.py (hunk 6 above)
Comment (yakkomajuri) [reply posthog/version_requirement.py:38]: Just checked on Cloud 👍 

good point though

File: posthog/apps.py (hunk 4 above)
Comment (yakkomajuri) [reply posthog/apps.py]: Renamed to service version requirement, agree with you here

File: posthog/settings.py (hunk 11 above)
Comment (yakkomajuri) [reply posthog/settings.py]: added

File: posthog/settings.py [hunk 12]
```diff
@@ -731,3 +732,13 @@ def show_toolbar(request):
 
 # keep in sync with plugin-server
 EVENTS_DEAD_LETTER_QUEUE_STATSD_METRIC = "events_added_to_dead_letter_queue"
+
+SERVICE_VERSION_REQUIREMENTS = [
+    ServiceVersionRequirement(service="postgresql", supported_version=">=11.0.0,<15.0.0",),
```
Comment (guidoiaquinti) [inline posthog/settings.py:737]: `<= 14.1` that I think is the latest version

File: posthog/settings.py [hunk 13]
```diff
@@ -731,3 +732,13 @@ def show_toolbar(request):
...
+    ServiceVersionRequirement(service="redis", supported_version=">=5.0.0,<7.0.0",),
```
Comment (guidoiaquinti) [inline posthog/settings.py:738]: Ditto: I think latest stable (against which we don't run any test afaik) is 6.2

File: posthog/settings.py (hunk 12 above)
Comment (yakkomajuri) [inline posthog/settings.py:737]:
```suggestion
    ServiceVersionRequirement(service="postgresql", supported_version=">=11.0.0,<=14.1.0",),
```

File: posthog/settings.py (hunk 13 above)
Comment (yakkomajuri) [inline posthog/settings.py:738]:
```suggestion
    ServiceVersionRequirement(service="redis", supported_version=">=5.0.0,<=6.2.0",),
```

File: posthog/settings.py [hunk 14]
```diff
@@ -731,3 +732,13 @@ def show_toolbar(request):
...
+]
+
+if PRIMARY_DB == AnalyticsDBMS.CLICKHOUSE:
+    SERVICE_VERSION_REQUIREMENTS = SERVICE_VERSION_REQUIREMENTS + [
+        ServiceVersionRequirement(service="clickhouse", supported_version=">=21.6.0,<21.7.0"),
```
Comment (guidoiaquinti) [inline posthog/settings.py:743]: This is a tricky one, I think cloud is where we run the oldest, Helm runs on `yandex/clickhouse-server:21.6.5`...

Comment (yakkomajuri) [reply posthog/settings.py:743]: Should be fine to support an entire minor range?

File: posthog/version_requirement.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+                f"service {service} cannot be used to specify a version requirement. service should be one of {services_str}"
+            )
+
+        self.service = service
+
+        try:
+            self.supported_version = SimpleSpec(supported_version)
+        except:
+            raise Exception(
+                f"supported_version is invalid. See the Docs for SimpleSpec: https://pypi.org/project/semantic-version/"
```
Comment (guidoiaquinti) [inline posthog/version_requirement.py:22]:
```suggestion
                f"the provided supported_version for service {service} is invalid. See the Docs for SimpleSpec: https://pypi.org/project/semantic-version/"
```

File: posthog/settings.py (hunk 14 above)
Comment (guidoiaquinti) [reply posthog/settings.py:743]: Sure, let's just make sure clouds runs on >=21.6 or this will fail

**Review by guidoiaquinti — APPROVED** [review]
LGTM

File: posthog/settings.py (hunk 14 above)
Comment (yakkomajuri) [reply posthog/settings.py:743]: Yup, Cloud is: `21.6.5.37`

**Comment by tiina303** [comment]
This might require using skip flag for the M1 special builds @mariusandra was doing, not sure what the version of clickhouse was there exactly.

**Pull request merged**